### ISSUE TITLE
[ILM] bucket listings consistently incomplete during short daily windows, interference from bucket cleaning presumed to be the cause

### ISSUE THREAD

**Issue by mikeatacme** [minio/minio#19257]
Expected Behavior:

Recursively listing objects in a Minio bucket should always return all the objects in the bucket.

Current Behavior:

Listing is unreliable for a brief amount of time.

Reason:

This speculatively seems to be connected with bucket cleaning.

Steps to reproduce:

This behavior was observed with a Minio bucket containing approximately 200,000 items. This bucket has a lifetime policy which deletes old items after 10 days. It was observed that attempts to recursively list bucket contents were occasionally unreliable. Further investigation revealed that the only unreliable time was the period immediately after midnight GMT.

Here's an excerpt from a long-running process which simply recursively lists the contents of a bucket every few minutes:

Mar 03 23:55:59 ----- bucketlister[1340904]: {"level":"info","bucket":"-----","seconds":9.711815076,"count":200551}
Mar 03 23:59:09 ----- bucketlister[1340904]: {"level":"info","bucket":"-----","seconds":9.171164849,"count":200637}
Mar 04 00:02:09 ----- bucketlister[1340904]: {"level":"info","bucket":"-----","seconds":0.103213523,"count":931}
Mar 04 00:05:10 ----- bucketlister[1340904]: {"level":"info","bucket":"-----","seconds":0.072882764,"count":931}
Mar 04 00:08:10 ----- bucketlister[1340904]: {"level":"info","bucket":"-----","seconds":0.074724009,"count":932}
Mar 04 00:11:10 ----- bucketlister[1340904]: {"level":"info","bucket":"-----","seconds":0.071902936,"count":932}
Mar 04 00:14:10 ----- bucketlister[1340904]: {"level":"info","bucket":"-----","seconds":0.072460442,"count":932}
Mar 04 00:17:11 ----- bucketlister[1340904]: {"level":"info","bucket":"-----","seconds":0.072869514,"count":932}
Mar 04 00:20:11 ----- bucketlister[1340904]: {"level":"info","bucket":"-----","seconds":0.072874116,"count":932}
Mar 04 00:23:11 ----- bucketlister[1340904]: {"level":"info","bucket":"-----","seconds":0.104411599,"count":932}
Mar 04 00:26:12 ----- bucketlister[1340904]: {"level":"info","bucket":"-----","seconds":0.071279461,"count":955}
Mar 04 00:29:26 ----- bucketlister[1340904]: {"level":"info","bucket":"-----","seconds":14.645729818,"count":184270}
Mar 04 00:32:37 ----- bucketlister[1340904]: {"level":"info","bucket":"-----","seconds":10.646592049,"count":184326}

I captured the trace with mc admin trace, but did not see anything revelatory.

Context:

During this window, some of our servers could potentially experience anomalous startup, getting only a fraction of the expected data necessary to begin an instance.

Our Environment:

Standalone Minio instance

Versions:
minio version RELEASE.2024-03-05T04-48-44Z (commit-id=1b5f28e99b145b0e3a5d99d1801a9b5cbfbdb091)
Runtime: go1.21.7 linux/amd64

Linux ----- 6.1.0-17-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.69-1 (2023-12-30) x86_64 GNU/Linux

**Comment by harshavardhana** [comment]
ILM by nature is asynchronous, and Listing by nature is atomic only up to 1000 entries. This is expected behavior. If you don't want ILM coming in the way, don't use it for prefixes that are mutating in nature, and you want the outputs for further processing. 

You cannot expect us to hold on to these expired objects.

**Comment by vcabbage** [comment]
@harshavardhana Minio is failing to list objects that are not expired. We're not expecting expired objects to be held on to.

EDIT to add additional context:

Using the example provided, only ~15k objects are being expired, yet Minio stops listing ~199k objects while ILM is running, then they show up again.

(I work with @mikeatacme)

**Comment by harshavardhana** [comment]
@vcabbage then we can definitely take a look.

What is the ILM on the bucket can you share that and also your setup size and number of objects.

Please provide a namespace dump. That will also help.

**Comment by vcabbage** [comment]
Our setup is just a single standalone server.

```
$ mc ilm rule ls --json local/wxalerts
{
 "status": "success",
 "target": "local/wxalerts",
 "config": {
  "Rules": [
   {
    "Expiration": {
     "Days": 10
    },
    "ID": "wxalerts",
    "Status": "Enabled"
   }
  ]
 },
 "updatedAt": "2024-02-15T21:28:11Z"
}
```

```
# mc du local/wxalerts
28GiB   204754 objects  wxalerts

# mc admin info local
●  127.0.0.1
   Uptime: 1 week
   Version: 2024-03-05T04:48:44Z
   Network: 1/1 OK
   Drives: 1/1 OK
   Pool: 1

Pools:
   1st, Erasure sets: 1, Drives per erasure set: 1

172 GiB Used, 11 Buckets, 848,757 Objects
1 drive online, 0 drives offline
```

> Please provide a namespace dump.

I'm not sure what a namespace dump is, can you let me know what I need to provide?

Thanks!

**Comment by harshavardhana** [comment]
> # mc du local/wxalerts
> 28GiB   204754 objects  wxalerts

Can you provide `mc ls -r --json` output of the entire namespace @vcabbage

**Comment by vcabbage** [comment]
Here's a link to the listing. It's about 250MB. https://s3.wasabisys.com/aa-share/minio-ls.json

**Comment by harshavardhana** [comment]
> Here's a link to the listing. It's about 250MB. https://s3.wasabisys.com/aa-share/minio-ls.json

and how does it change @vcabbage ? do you have objects that expire among this list on a daily basis?

**Comment by vadmeste** [comment]
@vcabbage also can you share the script that you do for listing objects ?

**Comment by vadmeste** [comment]
I think I am able to reproduce. I will update shortly.

**Comment by harshavardhana** [comment]
fixed via #19290
